lib/host/wasi/wasifunc.cpp

```diff
@@ -1847,5 +1847,145 @@ Expect<uint32_t> WasiSockShutdown::body(Runtime::Instance::MemoryInstance *,
+  /// Check memory instance from module.
+  if (MemInst == nullptr) {
+    return __WASI_ERRNO_FAULT;
+  }
+  const char *Node = nullptr;
+  if (NodeLen != 0) {
+    Node = MemInst->getPointer<const char *>(NodePtr, NodeLen);
+  }
+  const char *Service = nullptr;
+  if (ServiceLen != 0) {
+    Service = MemInst->getPointer<const char *>(ServicePtr, ServiceLen);
+  }
+  // service and node can not be nullptr at the same time
+  if (Service == nullptr && Node == nullptr) {
+    return __WASI_ERRNO_FAULT;
+  }
+  if (unlikely(MaxResLength > WASI::addrinfoArrayMax)) {
+    return __WASI_ERRNO_INVAL;
+  }
+  auto *const Hint = MemInst->getPointer<const struct __wasi_addrinfo_t *>(
+      HintsPtr, sizeof(struct __wasi_addrinfo_t));
+  auto *const ResLength = MemInst->getPointer<__wasi_size_t *>(
+      ResLengthPtr, sizeof(__wasi_size_t *));
+
+  struct addrinfo *TmpAddrinfoRes = nullptr;
+  if (auto Res =
+          Env.getAddrInfo(Node, Service, Hint, &TmpAddrinfoRes, ResLength);
+      unlikely(!Res)) {
+    return Res.error();
+  }
+  if (*ResLength > MaxResLength)
+    *ResLength = MaxResLength;
+
+  struct addrinfo *AddrinfoItem = TmpAddrinfoRes;
+  auto *const Res = MemInst->getPointer<uint8_t_ptr *>(ResPtr, sizeof(uint8_t));
+  auto *RealResItem = MemInst->getPointer<struct __wasi_addrinfo_t *>(
+      *Res, sizeof(struct __wasi_addrinfo_t));
+
+  for (size_t i = 0; i < *ResLength; i++) {
+    RealResItem->ai_flags =
+        static_cast<__wasi_aiflags_t>(AddrinfoItem->ai_flags);
+    RealResItem->ai_socktype =
+        static_cast<__wasi_sock_type_t>(AddrinfoItem->ai_socktype);
+    RealResItem->ai_protocol =
+        static_cast<__wasi_protocol_t>(AddrinfoItem->ai_protocol);
+    RealResItem->ai_family =
+        static_cast<__wasi_address_family_t>(AddrinfoItem->ai_family);
+
+    // process socket address
+    RealResItem->ai_addrlen = AddrinfoItem->ai_addrlen;
+    if (RealResItem->ai_addrlen) {
+      auto *SockeAddress = MemInst->getPointer<struct __wasi_sockaddr_t *>(
+          RealResItem->ai_addr, sizeof(struct __wasi_sockaddr_t));
+      switch (AddrinfoItem->ai_addr->sa_family) {
+      case AF_INET:
+        SockeAddress->sa_family = __WASI_ADDRESS_FAMILY_INET4;
+        break;
+      case AF_INET6:
+        SockeAddress->sa_family = __WASI_ADDRESS_FAMILY_INET6;
+        break;
+      }
+      // process sa_data in socket address
+      SockeAddress->sa_data_len = 0;
+      if (strlen(AddrinfoItem->ai_addr->sa_data) != 0) {
+        auto *SaDataBuff =
+            MemInst->getPointer<char *>(SockeAddress->sa_data, 14);
+        memcpy(SaDataBuff, TmpAddrinfoRes->ai_addr->sa_data, 14);
+        SockeAddress->sa_data_len = 14;
+      }
+    }
+
+    // process ai_canonname in addrinfo
+    if (AddrinfoItem->ai_canonname != NULL) {
+      RealResItem->ai_canonname_len = strlen(AddrinfoItem->ai_canonname);
+      if (RealResItem->ai_canonname_len + 1 > WASI::canonnameMaxSize) {
```

Comment:
Is the `canonnameMaxSize` defined in spec?

Reply:
no.For pre-allocating the `struct addrinfo` 's element --- `ai_canonname`, I define `WASI::canonnameMaxSize` .This is not defined in spec